library/general/src/lib/yast2/equatable.rb

```diff
@@ -0,0 +1,104 @@
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+# more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE LLC about this file by physical or electronic mail, you may
+# find current contact information at www.suse.com.
+
+module Yast2
+  # Mixin for classes that require to define a custom comparison
+  #
+  # By default, the methods #==, #eql? and #equal? in Object subclasses return true if both objects
+  # are the same (point to the same object in memory). Actually #eql? should return true if both
+  # objects refer to the same hash key. But in practice, two objects have the same hash key if they
+  # are the very same object. There are some exceptions like String, which returns the same hash key
+  # if they have the same value.
+  #
+  # The #eql? and #hash methods must be related. That is, if two objects are #eql?, then they should
+  # have the same #hash. This is important, otherwise we could have unexpected results in certain
+  # operations like subtracting Arrays. When performing the difference of two Arrays, the method
+  # used for comparing the objects in the Array depends on the Array length (see source code of
+  # Array#difference). If both Arrays have more than SMALL_ARRAY_LEN (i.e., 16) elements, then
+  # the #hash method is used. Otherwise it uses #eql?. This is one of the reason why #eql? and #hash
+  # should be paired.
+  #
+  # @example
+  #   class Foo
+  #     include Equatable
+  #
+  #     attr_reader :attr1, :attr2
+  #
+  #     eql_attr :attr1
+  #
+  #     def initialize(attr1, attr2)
+  #       @attr1 = attr1
+  #       @attr2 = attr2
+  #     end
+  #   end
+  #
+  #   foo1 = Foo.new("a", "b")
+  #   foo2 = Foo.new("a", "c")
+  #
+  #   foo1 == foo2      #=> true
+  #   foo1.eql?(foo2)   #=> true
+  #   foo1.equal?(foo2) #=> false
+  module Equatable
+    def self.included(base)
+      base.extend(ClassMethods)
+    end
+
+    # Class methods for defining the attributes to consider when comparing objects
+    module ClassMethods
+      # Inherited classes must remember the attributes for comparison from its parent class
+      def inherited(subclass)
+        subclass.eql_attr(*eql_attrs)
+      end
+
+      # Name of the attributes to consider when comparing objects
+      #
+      # @return [Array<Symbol>]
+      def eql_attrs
+        @eql_attrs || []
+      end
+
+      # Saves the name of the attributes to use when comparing objects
+      #
+      # @param names [Array<Symbol>]
+      def eql_attr(*names)
+        @eql_attrs ||= []
+        @eql_attrs += names
+      end
+    end
+
+    # Hash key to identify objects
+    #
+    # Objects with the same values for their eql_attrs have the same hash
+    #
+    # @return [Integer]
+    def hash
```

Comment:
A possible improvement is to have a `#eql_hash` method like this:

~~~
def eql_hash
  ([[:class, self.class]] + self.class.eql_attrs.map { |m| [m, send(m)] }).to_h
end

def hash
  eql_hash.hash
end
~~~

This would allow to classes that include the mixing to fine tune its hash (i.e, its comparison). For example, imagine we have a class that contains an array attribute, and we don't care about the elements order:

~~~
class MyClass
  include Yast2::Equatable

  attr_reader :array
  eql_attr :array

  def eql_hash
    h = super
    h[:array] = array.sort_by(&:hash) if h.keys.include?(:array)
    h
  end
~~~